src/main/java/net/snowflake/ingest/streaming/OffsetTokenVerificationFunction.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2024 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.streaming;
+
+/**
+ * Interface to provide a custom offset verification logic. If specified, verification failures will
+ * be logged as warnings and reported to Snowflake. This interface could be used when there are
+ * certain assumption about the offset token behavior. Please reach out to Snowflake if you notice
+ * any unexpected behaviors.
+ *
+ * <p>Below is an example that verifies that all offset tokens are monotonically increasing numbers:
+ *
+ * <pre>
+ *     <code>
+ *      private static final OffsetTokenVerificationFunction offsetTokenVerificationFunction =
+ *       (prevBatchEndOffset, curBatchStartOffset, curBatchEndOffset, rowCount) -> {
+ *         boolean isMatch = true;
+ *
+ *         if (curBatchStartOffset != null) {
+ *           try {
+ *             long curStart = Long.parseLong(curBatchStartOffset);
+ *             long curEnd = Long.parseLong(curBatchEndOffset);
+ *
+ *             // We verify that the end_offset - start_offset + 1 = row_count
+ *             if (curEnd - curStart + 1 != rowCount) {
+ *               isMatch = false;
+ *             }
+ *
+ *             // We verify that start_offset_of_current_batch = end_offset_of_previous_batch+1
+ *             if (prevBatchEndOffset != null) {
+ *               long prevEnd = Long.parseLong(prevBatchEndOffset);
+ *               if (curStart != prevEnd + 1) {
+ *                 isMatch = false;
+ *               }
+ *             }
+ *           } catch (NumberFormatException ignored) {
+ *             // Do nothing since we can't compare the offset, or report a mismatch if number is expected
+ *           }
+ *         }
+ *
+ *         return isMatch;
+ *       };
+ *     </code>
+ * </pre>
+ */
+public interface OffsetTokenVerificationFunction {
+  /**
+   * @param prevBatchEndOffset end offset token of the previous batch
+   * @param curBatchStartOffset start offset token of the current batch
+   * @param curBatchEndOffset end offset token of the current batch
+   * @param rowCount number of rows in the current batch
+   * @return a boolean indicates whether the verification passed or not, if not, we will log a
+   *     warning and report it to SF
+   */
+  boolean verify(
+      String prevBatchEndOffset,
+      String curBatchStartOffset,
+      String curBatchEndOffset,
+      long rowCount);
+}
```

Comment:
I am thinking out loud here and I dont have good suggestion. 
But, why do we need an interface here? I cant think of any more implementations of this method. 

I was mainly thinking of using something similar to a Predicate<E> where users can provide their custom implementation as this is too tied to 4 things accepted in this method.

Reply:
Took a look and I'm not sure how you would use Predicate in this case, do you have an example? These 4 things are the only 4 things we track at batch level related to offset so I'm not sure what else you can do.

Reply:
Yeah, I think its okay for us to go with this.. It's too tightly coupled with rowBuffer to make it more generalized.